### PR TITLE
Merge release3 into release4

### PR DESCRIPTION
**Pull request by DavidJayakumar**
### JIRA link (if applicable) ###

https://tools.hmcts.net/jira/browse/EUI-6587

### Change description ###

Move the recent canges from release 3 to release 4

**Does this PR introduce a breaking change?** (check one with "x")

```
[ ] Yes
[x] No
```

### CODE REVIEW

File: src/app/components/routing/application-routing.component.ts [hunk 1]
```diff
@@ -48,6 +48,9 @@ export class ApplicationRoutingComponent implements OnInit {
       if (bookingFeatureToggle && AppUtils.isBookableAndJudicialRole(userDetails)) {
         return this.router.navigate([ApplicationRoutingComponent.bookingUrl]);
       }
+      if (bookingFeatureToggle && AppUtils.isBookableAndJudicialRole(userDetails)) {
+        return this.router.navigate([ApplicationRoutingComponent.bookingUrl]);
+      }
```
Comment (connorpgpmcelroy) [inline src/app/components/routing/application-routing.component.ts:53]: Is this not just duplicated code?

**Review by connorpgpmcelroy — CHANGES_REQUESTED** [review]
Have one reuqested change to remove duplicated code

File: test/e2e/features/step_definitions/workAccessBookingUI/workAccessPage.steps.js [hunk 2]
```diff
@@ -105,6 +110,17 @@ defineSupportCode(function ({ And, But, Given, Then, When }) {
...
+        const displayedBookingsCount = await workAccessPage.getExistingBooksingCount();
+        for (let i = 0; i < displayedBookingsCount; i++ ){
+            const bookingDetails = await workAccessPage.getBookingDetails(i);
+            expect(bookingDetails.location.length > 0, `Booking location at index ${i} is not displayed or empty`).to.be.true
+            expect(bookingDetails.fromDate.length > 0, `Booking from date at index ${i} is not displayed or empty`).to.be.true 
+            expect(bookingDetails.toDate.length > 0, `Booking to date at index ${i} is not displayed or empty`).to.be.true 
+        } 
+    });
+
+
```
Comment (MateiHMCTS) [inline test/e2e/features/step_definitions/workAccessBookingUI/workAccessPage.steps.js:123]: extra row

File: test/e2e/features/step_definitions/workallocation/common.steps.js [hunk 3]
```diff
@@ -327,28 +327,30 @@ defineSupportCode(function ({ And, But, Given, Then, When }) {
...
+            }
 
+            for (let i = 0; i < messages.length; i++) {
+                let matchingMsgs = await ArrayUtil.filter(actualmessages, async (msg) => msg.includes(messages[i].message));
+                if (messages[i].message !== "") {
+                    expect(matchingMsgs.length > 0, `expected "${messages[i].message}" to be included in ${JSON.stringify(actualmessages)}`).to.be.true
+                }
+            }  
+        });
+
```
Comment (MateiHMCTS) [inline test/e2e/features/step_definitions/workallocation/common.steps.js:353]: extra row

File: test/e2e/utils/request.js [hunk 4]
```diff
@@ -86,7 +100,8 @@ class Request {
 
      async put(reqpath, data, headers){
         try {
-         return await http.put(reqpath, data, this.getRequestConfig(headers));
+            const config = await this.getRequestConfig(headers);
```
Comment (MateiHMCTS) [inline test/e2e/utils/request.js:103]: not aligned

File: test/nodeMock/workAllocation/bookingsData.js [hunk 5]
```diff
@@ -6,12 +6,12 @@ class BookingsData{
...
         this.bookingResponse = [];
         this.allLocations = [];
-        // this.setUpBookings([
-        //     { appointmentId : "100"},
-        //     { appointmentId: "101" },
-        //     { appointmentId: "102" },
-        //     { appointmentId: "103" }
-        // ]);
+        this.setUpBookings([
+            { appointmentId : "100"},
```
Comment (MateiHMCTS) [inline test/nodeMock/workAllocation/bookingsData.js:10]: styling issue

File: test/ngIntegration/tests/stepDefinitions/workallocation/waCommon.steps.js [hunk 6]
```diff
@@ -139,7 +139,18 @@ defineSupportCode(function ({ And, But, Given, Then, When }) {
                     roleAssignment[attr] = roleAssignment[attr] === "true";
                 }
             })
-
+            if (roleKeys.includes('bookable')){
+                const valueTypeAndValue = roleAssignment['bookable'].split(",");
+                if (valueTypeAndValue.length > 0 && valueTypeAndValue[1] && valueTypeAndValue[1].includes('string')){
+                    roleAssignment['bookable'] = valueTypeAndValue[0]
+                }else{
```
Comment (MateiHMCTS) [inline test/ngIntegration/tests/stepDefinitions/workallocation/waCommon.steps.js:146]: styling issue

File: test/ngIntegration/tests/stepDefinitions/workallocation/wa2.steps.js [hunk 7]
```diff
@@ -294,27 +294,16 @@ defineSupportCode(function ({ And, But, Given, Then, When }) {
...
-                    expect(actualLocationsList.includes(expectedloc), `${service} location ${expectedloc} not present ${JSON.stringify(body)} `).to.be.true;
-                } 
-            }
+        const expectedBody = datatable.rowsHash();
+        
+        for (const key of Object.keys(expectedBody)){
+            expect(body[key],`body param ${key} does not match expected`).to.equal(expectedBody[key]);
         }
 
+
```
Comment (MateiHMCTS) [inline test/ngIntegration/tests/stepDefinitions/workallocation/wa2.steps.js:306]: unnecessary row

**Review by MateiHMCTS — APPROVED** [review]
Everything is great!
Just noticed a few styling issues. would appreciate if you can solve them whenever.

File: test/e2e/features/pageObjects/caseEditPage.js [hunk 8]
```diff
@@ -355,9 +355,9 @@ class caseEditPage {
             case "DateTime":
                 var fieldDate = new Date(field.value).toLocaleString('en-UK', { hour12: true });
                 let time = fieldDate.split(", ")
-                var date1 = new Date(fieldDate);
-                date1.setDate(date1.getDate());
-                date1 = date1.getDate() + " " + monthNames[date1.getMonth()] + " " + date1.getFullYear() + ", " + time[1];
+                var date1 = time[0].split("/");
+                // date1.setDate(date1.getDate());
```
[the remark this answers is not in the document]
Comment (DavidJayakumar) [reply test/e2e/features/pageObjects/caseEditPage.js:359]: It is functional test. So I will leave it for Sreekanth

File: test/e2e/features/pageObjects/createCaseStartPage.js [hunk 9]
```diff
@@ -24,15 +24,43 @@ class CreateCaseStartPage {
...
+      } else {
+        locatorString += ` or contains(text(), '${option.trim()}')`;
+      }
+      i++;
+    }
+    locatorString = locatorString + ']';
+    
+    var e = element(by.xpath(locatorString));
     await BrowserWaits.waitForElement(e);
    await e.click();
```
[the remark this answers is not in the document]
Comment (DavidJayakumar) [reply test/e2e/features/pageObjects/createCaseStartPage.js:44]: fixed

**Pull request merged**